util/app/discovery/discovery.go

```diff
@@ -99,109 +90,167 @@ func DetectConfigManagementPlugin(ctx context.Context, appPath, repoPath, plugin
 	}).Debugf("pluginSockFilePath is: %s", pluginSockFilePath)
 
 	if pluginName != "" {
-		// check if the given plugin supports the repo
-		conn, cmpClient, connFound = cmpSupports(ctx, pluginSockFilePath, appPath, repoPath, fmt.Sprintf("%v.sock", pluginName), env, tarExcludedGlobs, true)
-		if !connFound {
+		c := newSocketCMPClientConstructorForPluginName(pluginSockFilePath, pluginName)
+		cmpClient, conn = namedCMPSupports(ctx, c, appPath, repoPath, env, tarExcludedGlobs)
+		if cmpClient == nil {
 			return nil, nil, fmt.Errorf("couldn't find cmp-server plugin with name %q supporting the given repository", pluginName)
 		}
 	} else {
 		fileList, err := os.ReadDir(pluginSockFilePath)
 		if err != nil {
-			return nil, nil, fmt.Errorf("Failed to list all plugins in dir, error=%w", err)
+			return nil, nil, fmt.Errorf("failed to list all plugins in dir: %w", err)
 		}
 		for _, file := range fileList {
 			if file.Type() == os.ModeSocket {
-				conn, cmpClient, connFound = cmpSupports(ctx, pluginSockFilePath, appPath, repoPath, file.Name(), env, tarExcludedGlobs, false)
-				if connFound {
+				c := newSocketCMPClientConstructorForPath(pluginSockFilePath, file.Name())
+				cmpClient, conn = unnamedCMPSupports(ctx, c, appPath, repoPath, env, tarExcludedGlobs)
+				if cmpClient != nil {
 					break
 				}
 			}
 		}
-		if !connFound {
+		if cmpClient == nil {
 			return nil, nil, fmt.Errorf("could not find plugin supporting the given repository")
 		}
 	}
 	return conn, cmpClient, nil
 }
 
-// matchRepositoryCMP will send the repoPath to the cmp-server. The cmp-server will
-// inspect the files and return true if the repo is supported for manifest generation.
-// Will return false otherwise.
-func matchRepositoryCMP(ctx context.Context, appPath, repoPath string, client pluginclient.ConfigManagementPluginServiceClient, env []string, tarExcludedGlobs []string) (bool, bool, error) {
-	matchRepoStream, err := client.MatchRepository(ctx, grpc_retry.Disable())
+// namedCMPSupports will return a client for the named plugin if it supports the given repository. It will also return
+// a closer for the connection.
+func namedCMPSupports(ctx context.Context, c CMPClientConstructor, appPath, repoPath string, env, tarExcludedGlobs []string) (pluginclient.ConfigManagementPluginServiceClient, io.Closer) {
```

Comment:
nitpick on the consistency of return types ordering, `DetectConfigManagementPlugin` return `io.Closer, pluginclient.ConfigManagementPluginServiceClient` but the newly introduced functions (`namedCMPSupports` , `unnamedCMPSupports`, and `getClientForFilepath`) return `pluginclient.ConfigManagementPluginServiceClient, io.Closer` .

Reply:
Updated `DetectConfigManagementPlugin`. I think `client, closer` is the more common convention in the code base.